Feature: Use MiniMessage formatting for lang file and implement a message cache

DaRacci opened the pull request:
I'm creating this PR in an early state to more so ask what the getFancyMessage method does so I can go about replacing due to the string being a component now.

Thatsmusic99 commented:
That's a good question, getFancyMessage is a method that uses a markdown-esque style to get commands and/or tooltips for a message, which I'm aware MiniMessage does largely replace but concerned about getting it converted fully with MiniMessage. For example:

\[[ACCEPT]](/tpayes Thatsmusic99) will run `/tpayes Thatsmusic99` when [ACCEPT] is clicked on.
\[Hover over me](Gotcha!) will show a tooltip saying "Gotcha!" when Hover over me is, well, hovered over.
\[Click here](https://www.youtube.com/watch?v=dQw4w9WgXcQ) will send people to the specified link when clicked.

It's a far less verbose system and still quite ingrained into some of the messages but not too many, if we do want to move over to MiniMessage though, I'm gonna need to determine 1. whether that requires unplugging earlier version support (I assume up to 1.15.2? I will admit I did this with my other plugin and it was the best decision I've ever made), 2. will it be shaded in? 3. Would this leave out Spigot servers? I know Paper is the next big thing and I'm all for it, but SpigotMC actively witch-hunts Paper exclusive plugins and you would be surprised how many people still use Spigot, I work customer support and see it. all. the. damn. time.

DaRacci commented:
As you can see in the current progress im using the spigot library to download minimessage and the bukkit platform, When you are on paper or a paper fork it wont do anything because that library is already present but when on spigot it will use the bukkit platform and newly download library to properly work. 
As for sorting out the fancy message stuff i have an understanding of its use now so ill see how i can replace it tomorrow.

Thatsmusic99 commented:
Ah yes, I do forget Spigot has that nifty library tool, that'll definitely save the worry. With versioning though, I assume older versions have to be dropped with these changes? I've already specified in AT's version page that 1.8-1.15.2 support will get dropped in the next major update (which was meant to be 5.6 but my plans did change), so if that's the case I can give a more solidified warning and see if I can work out something for those whose lives are basically dependent on keeping this plugin compatible for older versions.

DaRacci commented:
> Ah yes, I do forget Spigot has that nifty library tool, that'll definitely save the worry. With versioning though, I assume older versions have to be dropped with these changes? I've already specified in AT's version page that 1.8-1.15.2 support will get dropped in the next major update (which was meant to be 5.6 but my plans did change), so if that's the case I can give a more solidified warning and see if I can work out something for those whose lives are basically dependent on keeping this plugin compatible for older versions.

Yes so i will have to look into the compatibility of this since versions of paper from 1.16 - 1.17.1 are on adventure 4.9.* however this doesnt include minimessage which hopefully should allow me to use the most up to date minimessage, however actually i know that purpur includes minimessage 4.2.0-snapshot up to 1.18 so hmm ill really have to look at that.
Im hoping to avoid shading at all costs but i will have to see how i go with my testing.

Thatsmusic99 commented:
Yep, I try to avoid shading myself (excluding PaperLib and CM since those are made to be shaded), if you have a dependency that is version dependent, it's best to let the developer of that dependency update it rather than you do so.

Looking at you, Enjin.

Worst case scenario, if you're lost on what to do next, give me a yell and I'll see what I can do about it.

DaRacci commented:
> Hi DaRacci! I hope you're doing well - are you looking to continue with this pull request any time soon or do you still wish to have it on hold?

Hi, I'm currently on vacation so this is on hold for the time being, I'm still interested in picking this up when i am able too however.

Thatsmusic99 commented:
That's fine, thanks for catching up, I've been busy with my own work and exams :') but I hope you're having a good vacation, just wanted to check you still wanted to work on this!

DaRacci commented:
How do you feel on allowing some kotlin code in the codebase? I mainly work in kotlin and have some classes / utils i was going to clone over from my main plugin thats written in kotlin, if needed i can convert it to java however.

DaRacci commented:
I have rebased this ontop of my other PR so this will be good to go once that is

DaRacci commented:
It's also pretty much complete just doing some testing now

DaRacci commented:
Alright i think its all in working order so im gonna finish up on the refactoring PR so this is mergable.

Thatsmusic99 commented:
Thank you for the incredible work on this! I'm really looking forward to allowing MM to work with this upcoming version.

With regards to Kotlin, I have seen code in Kotlin and can vaguely read it but because I'm not proficient in it and would like to maintain at least some clarity for myself and those unfamiliar with it, would it be a major fuss to stick to Java?

If the libraries used are written in Kotlin I'm more than fine with that, because that's not AT's stuff.

Let me know when everything is ready for review. :)

DaRacci commented:
> Thank you for the incredible work on this! I'm really looking forward to allowing MM to work with this upcoming version.
> 
> With regards to Kotlin, I have seen code in Kotlin and can vaguely read it but because I'm not proficient in it and would like to maintain at least some clarity for myself and those unfamiliar with it, would it be a major fuss to stick to Java?
> 
> If the libraries used are written in Kotlin I'm more than fine with that, because that's not AT's stuff.
> 
> Let me know when everything is ready for review. :)

The only kotlin code I implemented was stuff which is part of my own library, I'm in the process of a giant refactor and module restructure which will allow me to once complete use the code i copied in as a library instead so you don't have to maintain it, If your fine with kotlin code for now while I complete that it would be much easier to wait a little for that and just not wanting to worry about maintaining it that's fine since the code i implemented is basically just a wrapper for the minimessage components and parsing the placeholders when needed.

Thatsmusic99 commented:
That sounds fine! Appreciate it.

Thatsmusic99 commented on AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/config/CustomMessages.java:
```diff
@@ -1,246 +1,312 @@
+        addExample("Teleport.eventBeforeTP_title.20", "<aqua>></aqua>&lTeleporting... &b<");
+        addExample("Teleport.eventBeforeTP_title.40", "<aqua>>></aqua>&lTeleporting... &b<<");
         addExample("Teleport.eventBeforeTP_title.60", "&b>>> &e&lTeleported! &b<<<");
         addExample("Teleport.eventBeforeTP_subtitle.0", "&bPlease do not move!");
```
Some of these aren't changed to MM format (e.g. &l to `<b>`), is there a reason for that?

Thatsmusic99 commented on AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/config/CustomMessages.java:
```diff
@@ -1,246 +1,312 @@
+            Click/hover over this text for more information.]
+            (&bCurrent Version » &7<version>|&bNew Version <dark_gray> &7{new-version}|&bTitle <dark_gray> &7<title>
+            |https://www.spigotmc.org/resources/advancedteleport.64139/)
+        """.trim());
```
Same for this but if it's not clear what the format intended is I can change that myself.

Thatsmusic99 commented on AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/config/CustomMessages.java:
```diff
@@ -501,12 +607,60 @@ public static void failable(
         @NotNull final String errorPath,
         @NotNull final BooleanSupplier isError,
-        final String... placeholders
+        final Object... placeholders
```
We can probably use TagResolvers instead of objects in this case so we can resort to altered display names and such.

Thatsmusic99 commented on AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/config/NewConfig.java:
```diff
@@ -557,13 +541,13 @@ It is recommend to avoid setting this option too high as this can have a worst c
         makeSectionLenient("death-management");
         addDefault("death-management.default", "bed;spawn");
-        addExample("death-management.world", "{default}");
+        addExample("death-management.world", "default");
```
Will need to account for changes like this to default options.

Thatsmusic99 reviewed (commented):
not just minimessage and message caching in this case, lol

Think there's just some things that need checking then we're good.

Thatsmusic99 commented:
Going to try and take some of this into my own hands since I do see some issues arising in the current v6 builds from using the Adventure library.

Thatsmusic99 commented:
Think this covers most of what I wanted to do. Some pretty big changes include:
- Using TagResolvers instead of pairs (by the looks of it, the lazy placeholders were just this with some extra steps)
- Re-add some form of the legacy formatting support AT had previously to backport and allow people to adjust to the new system at their own pace - the message caching already has the translated components. The plugin was made with the legacy stuff, gotta deal with it.
- Rebase for all of the changes on the target branch.

If there are any commits or changes you're not sure will work out, give me a shout. After that, I think I'll merge this and I can get v6 ready for beta use.

DaRacci commented on AdvancedTeleport-Bukkit/src/main/java/io/github/niestrat99/advancedteleport/api/AdvancedTeleportAPI.java:
```diff
@@ -154,7 +154,7 @@ public static boolean isWarpSet(@NotNull String name) {
 
                 // Call the post-create event to indicate success
-                new SpawnPostCreateEvent(spawn).callEvent();
+                Bukkit.getPluginManager().callEvent(new SpawnPostCreateEvent(spawn));
```
This could be reverted as the function was added to the CancellableATEvent class.

Thatsmusic99 replied:
Only issue with that is SpawnPostCreateEvent is not cancellable lmao, this event is used to pass the created Spawn object after it's been created and added to the database.

DaRacci commented:
> Think this covers most of what I wanted to do. Some pretty big changes include:
> 
>     * Using TagResolvers instead of pairs (by the looks of it, the lazy placeholders were just this with some extra steps)
> 
>     * Re-add some form of the legacy formatting support AT had previously to backport and allow people to adjust to the new system at their own pace - the message caching already has the translated components. The plugin was made with the legacy stuff, gotta deal with it.
> 
>     * Rebase for all of the changes on the target branch.
> 
> 
> If there are any commits or changes you're not sure will work out, give me a shout. After that, I think I'll merge this and I can get v6 ready for beta use.

I think that the way to go with regards to the legacy formatting is to explicitly say it will be removed in version X or to add an auto-conversion to the new formatting.

Besides that all looks good to me and i don't have anything else to add.

Thatsmusic99 commented:
I think I'll make the legacy transition a command for now, I at least have the code to translate everything and I haven't seen any major breakage thus far. I'll at least add a notice that it may be removed in a later version but nothing definitive on the chore branch. 

But thank you for confirming that with me, a lot of hard work went into this PR and I'm super excited to share this with everyone! Feel free to reach out if you need anything at all or have any questions and I'll be happy to help. :D
